x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,252 @@
+      action_id: action.action_id ?? uuidV4(),
+      '@timestamp': action['@timestamp'] ?? new Date().toISOString(),
+    };
+    acc.push(actionDoc);
+    actions.push(actionDoc);
+
+    return acc;
+  }, []);
+
+  try {
+    const bulkCreateActionsResponse = (await esClient.bulk(
+      {
+        body: bulkCreateActionsBody,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    )) as unknown as estypes.BulkResponse;
+
+    const [errorItems, successItems] = partition(
+      bulkCreateActionsResponse.items,
+      (a) => a.create?.error
+    );
+
+    successItems.forEach((successItem) => {
+      auditLoggingService.writeCustomAuditLog({
+        message: `User created Fleet action [id=${successItem.create?._id}]`,
+      });
+    });
+
+    // log errored items
+    if (errorItems.length) {
+      appContextService.getLogger().error(new FleetActionsError(JSON.stringify(errorItems)));
+    }
+
+    // return success items
+    return successItems.map((item, i) => ({
+      id: item.create!._id as string,
+      action: actions[i],
+    }));
+  } catch (createBulkActionsError) {
+    throw new FleetActionsError(
+      `Error creating bulk actions: ${createBulkActionsError.message}`,
+      createBulkActionsError
+    );
+  }
+};
+
+export const getActionsByIds = async (
+  esClient: ElasticsearchClient,
+  actionIds: string[]
+): Promise<FleetAction[]> => {
+  try {
+    const getActionsResponse = await esClient.search({
+      index: AGENT_ACTIONS_INDEX,
+      query: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                action_id: actionIds,
+              },
+            },
+          ],
+        },
+      },
+    });
+    return getActionsResponse.hits?.hits?.map((hit) => hit._source ?? {}) as FleetAction[];
+  } catch (getActionsByIdError) {
+    throw new FleetActionsError(
+      `Error getting action: ${getActionsByIdError.message}`,
+      getActionsByIdError
+    );
+  }
+};
+
+export const getActionsWithKuery = async (
+  esClient: ElasticsearchClient,
+  kuery: string
+): Promise<{ actions: FleetAction[]; total: number }> => {
+  try {
+    const query = toElasticsearchQuery(fromKueryExpression(kuery));
+    const getActionSearchResponse = await esClient.search({
+      index: AGENT_ACTIONS_INDEX,
+      query,
+    });
+
+    const actions = getActionSearchResponse.hits.hits.reduce<FleetAction[]>((acc, hit) => {
+      if (hit._source) {
+        acc.push(hit._source as FleetAction);
+      }
+      return acc;
+    }, []);
+
+    return {
+      actions,
+      total: actions.length,
+    };
+  } catch (getActionSearchError) {
+    throw new FleetActionsError(
+      `Error getting actions with kuery: ${getActionSearchError.message}`,
+      getActionSearchError
+    );
+  }
+};
+
+export const getActionResultsByIds = async (
+  esClient: ElasticsearchClient,
+  actionIds: string[]
+): Promise<{ actionsResults: FleetActionResponse[]; total: number }> => {
```

Comment:
Should also probably use `items` property to hold the array of items found.

also, same comment as above: the return list of items should include all `actionIds` requested and whether or not they were found. so something like:

```typescript

Promise<Array<{
  id: string; // action id requested (from 'actionIds' argument)
  action: FleetAction | undefined; 
}>>

```

Reply:
7bd6c021b31f6571570bdcef9fcc504be761b1b3